DOCSP-24545 Fundamentals - Retrieve (#24)
mongodb/docs-csharp: source/fundamentals/crud/read-operations/retrieve.txt

.. _charp-retrieve:

=============
Retrieve Data
=============

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

> In this guide, you can learn how to retrieve data from your MongoDB
> collections using the ``Find()`` method.

Reviewer: Suggestion: List the name of the product you are documenting at the beginning of the article so people can see quickly if they are on the wrong page.

Issue: Focus on value proposition of page in the overview. Introduce technical details like method names later in the page.

```suggestion
In this guide, you can learn how to retrieve data from MongoDB
with the MongoDB Dotnet Driver.
```

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``sample_restaurants.restaurants`` collection
from the :atlas:`sample datasets </sample-data>` provided by Atlas. See the 
:ref:`<csharp-quickstart>` to learn how to create a free MongoDB cluster and 
load this sample data.

> .. literalinclude:: ../../../includes/code-examples/Restaurant.cs

Author: Good catch. Added a line introducing the class

   :language: csharp
   :copyable:
   :dedent:

.. _csharp-retrieve-find:

Find Operation
--------------

You can use the ``Find()`` method to retrieve documents from a collection. 
The ``Find()`` method takes a **query filter**, that specifies your

> search criteria, and returns all matching documents as an ``IFindFluent`` interface.

Author: I don't think pointing out the IFindFluent interface is all that important to the rest of the documentation on this page so I removed it. I think it will be better to describe it more in other pages we have planned in the future. 

I added a little more detail to describe the query filter. Let me know how you feel about the new wording as it's not exactly what you suggested. Query Filters can be made in numerous ways and we are writing a page to detail all of them so I just linked to that (not yet made) page.

> search criteria, and returns all matching documents as an ``IFindFluent`` interface.

Reviewer: Looks good!

Find One Document
~~~~~~~~~~~~~~~~~

To find a single document in a collection, pass a query filter that specifies the

> criteria of the document you want to find, then append the ``FirstOrDefault`` or 

Reviewer: Issue: I'm not sure "append the method" is correct. I think the correct word for this is "chain" (see [fluent interface](https://en.wikipedia.org/wiki/Fluent_interface) wiki for more info).

```suggestion
criteria of the document you want to find, then chain the ``FirstOrDefault`` or 
```

``FirstOrDefaultAsync`` method. If more than one document match the query

> filter these methods return the *first* matching document.

Reviewer: Question/Suggestion: What is the first matching document? What order is used if no sort criteria is specified? Could a link be provided to learn more about this at the end of the section?

Author: I changed the wording of this to hopefully make it a little more clear. I'm hesitant to go into too much more detail on the specifics of the FirstOrDefault functionality (especially talking about sorting and things of that nature) on this page since I think there are a lot of details to it that could muddy the focus of this page.
I think the answer to "What is the first matching document" is largely depending on the query as well, which makes it hard to detail here.

Author: I like that, I think it's a good middle ground of explanation

.. tabs::

   .. tab:: Asynchronous
      :tabid: find-one-async

      .. code-block:: csharp
         :copyable: true

         var restaurants = await _restaurantsCollection.Find(filter).FirstOrDefaultAsync();

   .. tab:: Synchronous
      :tabid: find-one-sync

      .. code-block:: csharp
         :copyable: true

         var restaurants = _restaurantsCollection.Find(filter).FirstOrDefault();

> To see a full example of using ``Find()`` to find a single document, see

Reviewer: Issue: Type of software construct should be stated after name of construct.

```suggestion
To see a full example of using the ``Find()`` method to find a single document, see
```

:ref:`csharp-retrieve-additional-information`.

Find Multiple Documents

> ~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: Issue: Avoid [stacked headings](https://www.mongodb.com/docs/meta/style-guide/style/titles-and-headings/#style-and-structure)

> Find Documents Using A Cursor

Reviewer: Issue: I'm not sure the reader can avoid using a cursor (I believe the "ToList" method [just iterates over all documents in a cursor and adds them to a list](https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/IAsyncCursor.cs#L380-L385), not comprehensive research but I believe this is how it works in other drivers we've looked at).

Consider rephrasing elements of this section to be more about iterating through retrieved documents rather than "Find documents using a cursor/list" (as everything is using a cursor).

Author: That's a good point with cursors that I wasn't aware of. 

I rephrased this section to focus more on iteration and the methods used, rather that making a distinction between cursor and list.

`````````````````````````````

A cursor is a mechanism that allows an application to iterate over database 
results while holding only a subset of them in memory at a given time. Cursors
are useful when your ``Find()`` command returns a large amount of documents.

To find multiple documents in a collection and put them in a ``Cursor``, pass a 
query filter to the ``Find()`` method that specifies the criteria of the documents 
you want to find, then append the ``ToCursor`` or ``ToCursorAsync`` method as follows:

.. tabs::

   .. tab:: Asynchronous
      :tabid: find-cursor-async

      .. code-block:: csharp
         :copyable: true

         var restaurants = await _restaurantsCollection.Find(filter).ToCursorAsync();

   .. tab:: Synchronous
      :tabid: find-cursor-sync

      .. code-block:: csharp
         :copyable: true

         var restaurants = _restaurantsCollection.Find(filter).ToCursor();

Find Documents Using A List

> ```````````````````````````

Reviewer: Issue: Avoid using [three levels of headings](https://www.mongodb.com/docs/meta/style-guide/style/titles-and-headings/#style-and-structure).

To find multiple documents in a collection using a list, pass a query filter 
to the ``Find()`` method that specifies the criteria of the documents you want 
to find, then append the ``ToList`` or ``ToListAsync`` method as follows:

.. tabs::

   .. tab:: Asynchronous
      :tabid: find-list-async

      .. code-block:: csharp

         var restaurants = await _restaurantsCollection.Find(filter).ToList();

   .. tab:: Synchronous
      :tabid: find-list-sync

      .. code-block:: csharp

         var restaurants = _restaurantsCollection.Find(filter).ToListAsync();

To see a full example of using ``Find()`` to find multiple documents, see
:ref:`csharp-retrieve-additional-information`.

Find All Documents
~~~~~~~~~~~~~~~~~~

To find all documents in a collection, pass an empty
``BsonDocument`` as a query filter. Then append the ``ToList`` or ``ToListAsync``
method to return the documents as a ``List`` object, or the ``ToCursor`` or ``ToCursorAsync``
method to return the documents as a ``Cursor``.

To see a fully runnable example of using ``Find()`` to find all documents, see 
:ref:`csharp-retrieve-additional-information`.

Modify Find Behavior
~~~~~~~~~~~~~~~~~~~~

> You can modify the behavior of ``Find()`` by passing
> in ``FindOptions``. If you don't specify any options, the driver uses the 

Reviewer: Issue: specify the type of the programming constructs you are referring to

```suggestion
You can modify the behavior of the ``Find()`` method by passing a
a ``FindOptions`` object. If you don't specify any options through this object, the driver uses the 
```

default values for each option.

You can configure the commonly used options with the following methods:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Method
     - Description

   * - ``BatchSize`` 
     - | Gets or sets the number of documents to return per batch.

   * - ``Collation`` 
     - | Sets the collation options. 

   * - ``Comment`` 
     - | Sets the comment to the query.

   * - ``Hint`` 
     - | Sets the hint for which index to use.
       | Default: ``0``

   * - ``MaxTime`` 

>      - | Sets the maximum execution time on the server for this operation.

Reviewer: Question/Possible Issue: Is there a reason only 1 option is given a default value?

Author: This is a good example of the copy/paste errors that we've been talking about lately. I meant to get rid of all the default values. Good Catch! I also removed the reference in the sentence above that talks about these default values.

To see a full list of available options, see 

> `FindOptions Properties <{+api-root+}/html/Properties_T_MongoDB_Driver_FindOptions.htm>`__.

Reviewer: Issue: I think this link [didn't work for me](https://mongodb.github.io/mongo-csharp-driver/2.17/apidocs/html/html/Properties_T_MongoDB_Driver_FindOptions.htm). Is it working on your end?

Author: There was a typo in the link, nice catch. Fixed

Example
~~~~~~~

> The following example passes a query filter and ``FindOptions`` to the ``Find()``
> method which performs the following actions:

Reviewer: Issue: I believe this description doesn't put the value proposition (configure/modify ``Find`` behavior) of the example at the beginning of the sentence.

```suggestion
The following example configures the behavior of the ``Find()`` method with a ``FindOptions`` object.
The example performs the following actions:
```

- Finds all documents with "Pizza" in the ``cuisine`` field
- Sets the ``BatchSize`` to ``3``
- Stores the results in a ``Cursor`` object
- Prints the number of documents currently held in the ``Cursor``

.. io-code-block::
   :copyable: true

   .. input::
      :language: csharp

      var filter = Builders<Restaurant>.Filter.Eq("cuisine", "Pizza");

      var findOptions = new FindOptions
      {
         BatchSize = 3
      };

      using (var cursor = _restaurantsCollection.Find(filter, findOptions).ToCursor())
      {
         cursor.MoveNext();
         WriteLine($"Number of documents in cursor: {cursor.Current.Count()}");
      }

   .. output::
      
      Number of documents in cursor: 3

> .. tip::

Reviewer: Suggestion: Add a title to the admonition to help with scannability

```suggestion
.. tip:: Clean Up
```

   Create a ``Cursor`` with a `using statement <https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-statement>`__ to

>    ensure the proper disposal of resources.

Reviewer: Issue: I don't know what is meant by "ensure the proper disposal of resources". Could this be clarified? What resources are disposed? What method does a "using statement" let you avoid calling?

Author: I tried to clarify this a little bit better. Let me know if this gets the point across. If not and if it seems too confusing or starts to get too cumbersome to try and describe here then I'm also open to the idea of removing it entirely. It may be more fitting in a future different page

Reviewer: I like this tip and hope you keep it. I like your update as well! Only potential tweak I'd make is to provide a link to the API documentation for `Dispose` at the end of the admonition.

``BsonDocument`` as a query filter. Then chain the ``ToList`` or ``ToListAsync``
method to return the documents as a ``List`` object, or the ``ToCursor`` or ``ToCursorAsync``
method to return the documents as a ``Cursor``.

To see a fully runnable example of using ``Find()`` to find all documents, see 
:ref:`csharp-retrieve-additional-information`.

Modify Find Behavior
~~~~~~~~~~~~~~~~~~~~

You can modify the behavior of the ``Find()`` method by passing
a ``FindOptions`` object.

You can configure the commonly used options with the following methods:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Method
     - Description

   * - ``BatchSize`` 
     - | Gets or sets the number of documents to return per batch.

   * - ``Collation`` 
     - | Sets the collation options. 

   * - ``Comment`` 
     - | Sets the comment to the query.

   * - ``Hint`` 
     - | Sets the hint for which index to use.

   * - ``MaxTime`` 
     - | Sets the maximum execution time on the server for this operation.

To see a full list of available options, see 
`FindOptions Properties <{+api-root+}/Properties_T_MongoDB_Driver_FindOptions.htm>`__.

Example

> ~~~~~~~

Reviewer: Sounds good. In that case, could the example the text introducing the example be rephrased to focus less on configuring the find method? I think the following could work:

```
The following example performs the these actions:

- Retrieves all documents with "Pizza" in the ``cuisine`` with the ``Find()`` method
- Configures the behavior of the ``Find()`` method with a ``FindOptions`` object
... rest stays the same...
```

The following example configures the behavior of the ``Find()`` method with a 
``FindOptions`` object. The example performs the following actions:

- Finds all documents with "Pizza" in the ``cuisine`` field
- Sets the ``BatchSize`` to ``3``

> - Stores the results in a ``Cursor`` object

Author: Changed all the references of ``Cursor`` to simply cursor, since I was intending to reference the cursor mechanism.

.. _charp-retrieve:

=============
Retrieve Data
=============

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to retrieve data from your MongoDB
collections with the {+driver-long+}.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``sample_restaurants.restaurants`` collection
from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
free MongoDB Atlas cluster and load the sample datasets, see the :ref:`<csharp-quickstart>`.

The following ``Restaurants`` class models the documents in this collection:

.. literalinclude:: ../../../includes/fundamentals/code-examples/crud/Retrieve.cs
   :language: csharp
   :copyable:
   :dedent:

.. _csharp-retrieve-find:

Find Documents
--------------

Use the ``Find()`` method to retrieve documents from a collection. 
The ``Find()`` method takes a **query filter** and returns all matching documents.
A query filter is an object that specifies the documents you want to retrieve in
your query. 

> To learn more about query filters, see :ref:`csharp-specify-query`.

Author: I'm currently working on this page as well, so it's not working yet but should be soon as the other page is in. I have a todo to make sure it works once specify-query is done.

If you are returning a small number of documents, or need your results returned
as a ``List`` object, use the ``ToList()`` or ``ToListAsync()`` methods.

To find multiple documents in a collection and hold them in memory as a list, pass a query filter 
to the ``Find()`` method that specifies the criteria of the documents you want 
to find, then chain the ``ToList()`` or ``ToListAsync()`` method as follows:

.. tabs::

   .. tab:: Asynchronous
      :tabid: find-list-async

      .. code-block:: csharp

         var restaurants = await _restaurantsCollection.Find(filter).ToList();

   .. tab:: Synchronous
      :tabid: find-list-sync

      .. code-block:: csharp

         var restaurants = _restaurantsCollection.Find(filter).ToListAsync();

To see a full example of using the ``Find()`` method to find multiple documents,
see :ref:`csharp-retrieve-additional-information`.

Find All Documents
~~~~~~~~~~~~~~~~~~

To find all documents in a collection, pass an empty
``BsonDocument`` as a query filter. Then chain the ``ToList()`` or ``ToListAsync()``
method to return the documents as a ``List`` object, or the ``ToCursor()`` or ``ToCursorAsync()``

> method to return the documents as a cursor.

Reviewer: Question/Possible Issue: Any reason ``ToList()`` is presented first here? I think trying to hold an entire collection in a list may not be a good thing to implicitly advocate for.

Suggestion: This section is short, and I think the second sentence of the first paragraph largely duplicates the preceding subsection and can be deleted. I think the following edits could be beneficial:

- Remove the Find All documents section and move the content to an admonition
- Place the admonition in the "Find Multiple Documents" section
- Shorten the content as follows:

```
.. note: Find All Documents

   To find all documents in a collection, pass an empty ``BsonDocument`` as a query filter.

   To see a fully runnable example of using the ``Find()`` method to find all documents, see 
   :ref:`csharp-retrieve-additional-information`.   
```

Author: Good suggestion

To see a fully runnable example of using the ``Find()`` method to find all documents, see 
:ref:`csharp-retrieve-additional-information`.

Modify Find Behavior
~~~~~~~~~~~~~~~~~~~~

You can modify the behavior of the ``Find()`` method by passing
a ``FindOptions`` object.

You can configure the commonly used options with the following methods:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Method
     - Description

>    * - ``BatchSize`` 

Reviewer: Question: Any reason this did not receive method formatting `()`?

Author: Adding `()` made it seem like you needed to call it like a method if you wanted to use it, so I left it since you pass it in an options object when using it.

>      - | Gets or sets the number of documents to return per batch.
> 
>    * - ``Collation`` 
>      - | Sets the collation options. 
> 
>    * - ``Comment`` 
>      - | Sets the comment to the query.
> 
>    * - ``Hint`` 
>      - | Sets the hint for which index to use.

Reviewer: Suggestion: These definitions are largely circular. I believe all four of the definitions highlighted define the option using the option name (not super helpful if you don't already know what the option name is referring to).

Consider providing an explanation of each option that helps give a reader something to relate the option to, or providing a link to let the reader learn more about what the option is doing:

```suggestion
     - | Gets or sets the number of documents hold in a cursor at a given time.

   * - ``Collation`` 
     - | Sets the collation options. To learn more about collations, see the TODO: collations page. 

   * - ``Comment`` 
     - | Sets the comment to the query. To learn more about query comments, see https://www.mongodb.com/docs/manual/reference/operator/query/comment/.

   * - ``Hint`` 
     - | Sets the hint for which index to use.
```

Author: Added all but the collation link since we don't have a plan of making a collations page in the MVP. Happy to add it when we do add that page

- Sets the ``BatchSize`` to ``3``
- Stores the results in a cursor
- Prints the number of documents currently held in the cursor

.. io-code-block::
   :copyable: true

   .. input::
      :language: csharp

      var filter = Builders<Restaurant>.Filter.Eq("cuisine", "Pizza");

      var findOptions = new FindOptions
      {
         BatchSize = 3
      };

      using (var cursor = _restaurantsCollection.Find(filter, findOptions).ToCursor())
      {
         cursor.MoveNext();
         WriteLine($"Number of documents in cursor: {cursor.Current.Count()}");
      }

   .. output::
      
      Number of documents in cursor: 3

.. tip:: Clean Up

   Create a cursor with a `using statement <https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-statement>`__ to
   automatically invoke the 
   `Dispose() <https://learn.microsoft.com/en-us/dotnet/api/system.idisposable.dispose?view=net-7.0>`__
   method once the cursor is no longer in use.

.. _csharp-retrieve-additional-information:

Additional Information
----------------------

.. TODO: Links to creating filter with Builders and LINQ

To learn more about query filters, see :ref:`csharp-specify-query`

> For fully runnable examples of the ``Find()`` method, see the 
> :ref:`Usage Examples <csharp-find-one>`.

Reviewer: Suggestion: Could the destination of this link be made more precise?
Issue: For -> "To view" to add imperative mood

```suggestion
To view a fully runnable example of the ``Find()`` method, see the 
:ref:`<csharp-find-one>` page.
```
